test/server/server_test.cc

```diff
@@ -568,17 +568,29 @@ TEST_P(ServerInstanceImplTest, BootstrapNodeWithoutAccessLog) {
                             "An admin access log path is required for a listening server.");
 }
 
+namespace {
+void bindAndListenTcpSocket(const Network::Address::InstanceConstSharedPtr& address,
+                            const Network::Socket::OptionsSharedPtr& options) {
+  auto socket = std::make_unique<Network::TcpListenSocket>(address, options, true);
+  // Some kernels erroneously allow `bind` without SO_REUSEPORT for addresses
+  // with some other socket already listening on it, see #7636.
+  if (::listen(socket->ioHandle().fd(), 1) != 0) {
+    throw EnvoyException(fmt::format("cannot listen:{}", strerror(errno)));
+  }
+}
+} // namespace
+
 // Test that `socket_options` field in an Admin proto is honored.
 TEST_P(ServerInstanceImplTest, BootstrapNodeWithSocketOptions) {
   ASSERT_NO_THROW(initialize("test/server/node_bootstrap_with_admin_socket_options.yaml"));
   const auto address = server_->admin().socket().localAddress();
-  EXPECT_THAT_THROWS_MESSAGE(std::make_unique<Network::TcpListenSocket>(address, nullptr, true),
-                             EnvoyException, HasSubstr("Address already in use"));
+  EXPECT_THAT_THROWS_MESSAGE(bindAndListenTcpSocket(address, nullptr), EnvoyException,
+                             HasSubstr("Address already in use"));
```

Comment:
Oh, one question -- Is the purpose of this change to allow the test to pass on when the kernel allows the bind without SO_REUSEPORT? Because it seems to me that in that case `bindAndListenTcpSocket` will throw an exception with "cannot listen" and the errno and the test will still fail.

Reply:
That all makes sense except the last part. It seems to me that the exception message will not be "Address already in use" and so the test will fail.

Reply:
I'll double check, but my expectation is that `strerror` should convert EADDRINUSE error number to "Address already in use" text

Reply:
Updated the code to verify `errno` directly + added comments about the intent.
Does it help?

Reply:
You're right, I was reading the original EXPECT_THROW_HAS_MESSAGE line incorrectly the whole time. It was fine. The new version is very explicit about the error number, which is probably good since I'm not sure those error messages are necessarily the same across operating systems.